File: python/verify_url.py
"""Live HTTP verify client for the Agent Disclosure Protocol.

Fetches a counterparty's signed disclosure from its well-known endpoint, then
runs the native verifier (signature + agentId<->publicKey binding) and prints a
transact/refuse decision. The native side of the live cross-process interop test
against the TS `verify-url` / `verifierService`.

    python verify_url.py <baseUrl> [--now <iso8601>]

Exit 0 when the disclosure is valid (would transact), 1 on any invalid
disclosure OR transport error. Never crashes on hostile input.
"""

import argparse
import sys
import urllib.error
import urllib.request
from datetime import datetime, timezone

from agent_disclosure import evaluate_raw
from agent_disclosure.verify import VerificationPolicy


def _now_iso() -> str:
    """Current UTC instant as an ISO-8601 string, the freshness clock for a live verify."""
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"

WELL_KNOWN_PATH = "/.well-known/agent-disclosure"


def fetch_disclosure(base_url: str, timeout: float = 10.0) -> str:
    """GET <base>/.well-known/agent-disclosure and return the raw body text."""
    url = base_url.rstrip("/") + WELL_KNOWN_PATH
    req = urllib.request.Request(url, headers={"Accept": "application/json"})
    with urllib.request.urlopen(req, timeout=timeout) as resp:  # noqa: S310 — operator-supplied URL
        charset = resp.headers.get_content_charset() or "utf-8"
        return resp.read().decode(charset)


def main(argv=None) -> int:
    parser = argparse.ArgumentParser(
        prog="verify_url.py",
        description="Fetch and verify a counterparty's Agent Disclosure over HTTP.",
    )
    parser.add_argument("base_url", help="Base URL hosting /.well-known/agent-disclosure")
    parser.add_argument(
        "--now",
        default=None,
        help="ISO-8601 instant for the freshness window (default: current UTC time).",
    )
    parser.add_argument(
        "--timeout", type=float, default=10.0, help="HTTP timeout in seconds (default: 10)."
    )
    args = parser.parse_args(argv)

    try:
        raw = fetch_disclosure(args.base_url, timeout=args.timeout)
    except (urllib.error.URLError, OSError, ValueError) as e:
        print(f"decision: refuse")
        print(f"transport error: {e}")
        return 1

    policy = VerificationPolicy(now=args.now or _now_iso())
    verdict = evaluate_raw(raw, policy)

    print(f"decision: {verdict.decision}")
    if verdict.failed:
        print(f"failed checks: {', '.join(verdict.failed)}")
    for reason in verdict.reasons:
        print(f"  - {reason}")

    return 0 if verdict.decision == "transact" else 1


if __name__ == "__main__":
    sys.exit(main())

File: rust/src/bin/verify-url.rs
#![allow(clippy::result_large_err)]
//! `verify-url` — fetch a counterparty's published disclosure over HTTP and verify it.
//!
//! Usage: `verify-url <base-url>`
//!
//! Issues an HTTP GET to `<base-url>/.well-known/agent-disclosure`, parses the signed
//! envelope, and runs the core acceptance check: the ed25519 signature over the
//! canonical disclosure plus the identity binding `agentId == signature.publicKey`.
//! Prints the decision and exits 0 when the disclosure is valid, 1 on any
//! invalid-or-transport error (unreachable host, non-200, malformed body, bad binding,
//! signature mismatch). Policy evaluation (freshness, red-team, etc.) is out of scope
//! here — this is the wire-level "is this a genuinely signed, correctly-bound
//! disclosure" gate.

use agent_disclosure::{verify_disclosure_signature, SignedDisclosure};
use serde_json::Value;

const WELL_KNOWN_PATH: &str = "/.well-known/agent-disclosure";

fn main() {
    let mut args = std::env::args().skip(1);
    let base = match args.next() {
        Some(b) => b,
        None => {
            eprintln!("usage: verify-url <base-url>");
            std::process::exit(1);
        }
    };

    match run(&base) {
        Ok(report) => {
            println!("{report}");
            std::process::exit(0);
        }
        Err(reason) => {
            eprintln!("INVALID: {reason}");
            std::process::exit(1);
        }
    }
}

fn run(base: &str) -> Result<String, String> {
    let url = format!("{}{}", base.trim_end_matches('/'), WELL_KNOWN_PATH);

    let body = ureq::get(&url)
        .call()
        .map_err(|e| format!("transport error fetching {url}: {e}"))?
        .into_string()
        .map_err(|e| format!("could not read response body: {e}"))?;

    let value: Value =
        serde_json::from_str(&body).map_err(|e| format!("response is not valid JSON: {e}"))?;

    let signed = SignedDisclosure::from_value(&value)?;
    verify_disclosure_signature(&signed)?;

    let agent_id = value
        .pointer("/disclosure/agentId")
        .and_then(|v| v.as_str())
        .unwrap_or("<unknown>");
    let disclosure_id = value
        .pointer("/disclosure/disclosureId")
        .and_then(|v| v.as_str())
        .unwrap_or("<unknown>");

    Ok(format!(
        "VALID: disclosure {disclosure_id} from agent {agent_id} (signature + identity binding verified) at {url}"
    ))
}
